monkey/agent_plugins/exploiters/smb/src/smb_client.py

```diff
@@ -0,0 +1,307 @@
+        self._authenticated = False
+
+    def connected(self) -> bool:
+        return self._authenticated
+
+    def connect_with_user(self, host: TargetHost, credentials: Credentials, timeout: float):
+        """
+        Connect to target host over SMB
+
+        :param host: A target host to which to connect
+        :param credentials: Credentials to use when connecting
+        :param timeout: SMB connection timeout
+        :raise Exception: If connection fails
+        """
+        self._create_smb_connection(host)
+        self._smb_login(credentials)
+        self.set_timeout(timeout)
+        if self._logout_guest():
+            raise Exception("Logged in as guest")
+
+    def _get_smb_connection(self) -> SMBConnection:
+        if not self._smb_connection:
+            raise Exception("SMB connection not established")
+        return self._smb_connection
+
+    def _create_smb_connection(self, host: TargetHost):
+        """Connect to host over SMB. Raise Exception if connection fails"""
+        try:
+            self._smb_connection = SMBConnection(
+                str(host.ip), str(host.ip), sess_port=445, preferredDialect=SMB_DIALECT
+            )
+            return
+        except SessionError as err:
+            logger.debug(f"Failed to create SMB connection to {host} on port 445: {err}")
+
+        try:
+            # "*SMBSERVER" and port 139 is a special case. See doc for SMBConnection
+            self._smb_connection = SMBConnection("*SMBSEVER", str(host.ip), sess_port=139)
+            return
+        except SessionError as err:
+            logger.debug(f"Failed to create SMB connection to {host} on port 139: {err}")
+
+        raise Exception(f"Failed to create SMB connection to {host}")
+
+    def _smb_login(self, credentials: Credentials):
+        """Raise SessionError if login fails"""
+
+        self._get_smb_connection().login(
+            user=credentials.identity.username,
+            domain="",
+            **self._build_args_for_secrets(credentials),
+        )
+        self._authenticated = True
+        self._authenticated_credentials = self._get_smb_connection().getCredentials()
+
+    @staticmethod
+    def _build_args_for_secrets(credentials: Credentials) -> Dict[str, str]:
+        args = {"password": ""}
+
+        if isinstance(credentials.secret, Password):
+            secret_type = "password"
+        elif isinstance(credentials.secret, LMHash):
+            secret_type = "lmhash"
+        elif isinstance(credentials.secret, NTHash):
+            secret_type = "nthash"
+        else:
+            return args
+
+        args.update({secret_type: get_plaintext_secret(credentials)})
+        return args
+
+    def _logout_guest(self):
+        """Return True if logged in as guest. Raise SessionError if logout fails"""
+        smb_connection = self._get_smb_connection()
+        if smb_connection.isGuestSession() > 0:
+            smb_connection.logoff()
```

Comment:
NIT: Checking if logged in as guest and logging off are separate actions, method should only do one thing. This method should probably just check if `_logged_in_as_guest`

Reply:
Either that or we can define the method to be idempotent and raise an error if logout fails.